sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_local.asciidoc

```diff
@@ -0,0 +1,312 @@
+local memory capacity (as reported by `local_mem_size`) then the implementation
+must issue a diagnostic.
+
+[source,c++]
+----
+operator T&() const noexcept;
+----
+_Returns_: A reference to the object stored in the work-group local memory
+associated with this instance of `work_group_local`.
+
+[source,c++]
+----
+const work_group_local<T>& operator=(const T& value) const noexcept;
+----
+_Constraints_: Available only if `std::is_array_v<T>>` is false.
+
+_Effects_: Replaces the value referenced by `*ptr` with `value`.
+
+_Returns_: A reference to this instance of `work_group_local`.
+
+[source,c++]
+----
+T* operator&() const noexcept;
+----
+_Returns_: A pointer to the work-group local memory associated with this
+instance of `work_group_local` (i.e., `ptr`).
+
+
+==== Kernel properties
+
+The `work_group_local_size` property must be passed to a kernel to determine
+the run-time size of the work-group local memory allocation associated with
+all `work_group_local` variables of unbounded array type.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+struct work_group_local_size {
+  constexpr work_group_local_size(size_t bytes) : value(bytes) {}
+  size_t value;
+}; // work_group_local_size
+
+using work_group_local_size_key = work_group_local_size;
+
+template <>struct is_property_key<work_group_local_size_key> : std::true_type {};
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+|===
+|Property|Description
+
+|`work_group_local_size`
+|The `work_group_local_size` property describes the amount of dynamic
+work-group local memory required by the kernel in bytes.
+
+|===
+
+
+==== Usage examples
+
+===== Allocations with size known at compile-time
+
+[source,c++]
+----
+using namespace syclex = sycl::ext::oneapi::experimental;
+
+/* optional: static const */ syclex::work_group_local<int> program_scope_scalar;
+/* optional: static const */ syclex::work_group_local<int[16]> program_scope_array;
+
+void foo() {
+  /* optional: static const */ syclex::work_group_local<int> function_scope_scalar;
+  function_scope_scalar = 1; // assignment via overloaded = operator
+  function_scope_scalar += 2; // += operator via implicit conversion to int&
+  int* ptr = &function_scope_scalar; // conversion to pointer via overloaded & operator
+}
+
+void bar() {
+  /* optional: static const */ sylex::work_group_local<int[64]> function_scope_array;
+  function_scope_array[0] = 1; // [] operator via implicit conversion to int(&)[64]
+  int* ptr = function_scope_array; // conversion to pointer via implicit conversion to int(&)[64]
+}
+----
+
+===== Allocations with size unknown at compile-time
+
+[source,c++]
+----
+using namespace syclex = sycl::ext::oneapi::experimental;
+
+/* optional: static const */ syclex::work_group_local<int[]> dynamic_program_scope_array;
+
+...
+
+q.parallel_for(sycl::nd_range<1>{N, M},
+  syclex::properties{syclex::work_group_local_size(1024)},
```

Comment:
```suggestion
  syclex::properties{syclex::work_group_local_size(1024*sizeof(int))},
```

I would do this to make it clear that the allocation size is in bytes.

Reply:
Good idea.  I also swapped the 1024 for `M`, since the whole point of this example is to use a dynamic value.  See 407fac11d3e945c5f1bd3f22373dfefd66e5c44b.